### PR TITLE
Fix Tensorflow Golden Container Smoke test

### PR DESCRIPTION
**Pull request by rikenm1**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] If you are adding/removing a .spec file that has multiple-versions supported, please add [@microsoft/cbl-mariner-multi-package-reviewers](https://github.com/orgs/microsoft/teams/cbl-mariner-multi-package-reviewers) team as reviewer [(Eg. golang has 2 versions 1.18, 1.21+)](https://github.com/microsoft/azurelinux/tree/2.0/SPECS/golang)
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
Tensorflow Golden container smoke test is failing.
###### Change Log  

- Change: Update Keras to 3.3.3 and add a patch to enable wheel installation. Also use keras build to enable keras API for tensorflow
- Change: Remove D_GLIBCXX_ASSERTIONS from optree build 

###### Does this affect the toolchain?  

NO

###### Test Methodology

- Pipeline build id: [605833](https://dev.azure.com/mariner-org/mariner/_build/results?buildId=605833&view=results)

### CODE REVIEW

File: SPECS/keras/0001-Add-Keras-3.3.3.patch [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
++root_path = ""
+ to_copy = ["setup.py", "README.md"]
+ 
+ 
+@@ -86,7 +87,8 @@
+         move_tf_keras_directory()
+         print(os.getcwd())
+ 
+-        from keras.src.version import __version__  # noqa: E402
++        #from keras.src.version import __version__  # noqa: E402
```
Comment (rikenm1) [inline SPECS/keras/0001-Add-Keras-3.3.3.patch:17]: was getting an error so just hardcoded this instead

Comment (jslobodzian) [reply SPECS/keras/0001-Add-Keras-3.3.3.patch:17]: What was the error?  Can you comment in the SPEC file.  And if we absolutely need this, can you add something to the %build section that fails the build when somebody upgrades keras to anything other than 3.3.3?

File: SPECS/python-optree/python-optree.spec [hunk 2]
```diff
@@ -30,18 +32,28 @@ A PyTree is a recursive structure that can be an arbitrarily nested Python conta
 
 
 %build
+
+export FFLAGS=$(echo $FFLAGS | sed 's/\(-Wp,-D_GLIBCXX_ASSERTIONS\)//g')
+export CXXFLAGS=$(echo $CXXFLAGS | sed 's/\(-Wp,-D_GLIBCXX_ASSERTIONS\)//g')
+export CFLAGS=$(echo $CFLAGS | sed 's/\(-Wp,-D_GLIBCXX_ASSERTIONS\)//g')
+export RPM_OPT_FLAGS=$(echo $RPM_OPT_FLAGS | sed 's/\(-Wp,-D_GLIBCXX_ASSERTIONS\)//g')
```
Comment (jslobodzian) [inline SPECS/python-optree/python-optree.spec:39]: Likewise for RPM_OPT_FLAGS and FCFLAGS.  I think it is safe to limit the change to just the CXXFLAGS or CFLAGS.  I believe it's using gcc not g++ to compile, so probably the CFLAGS are the only ones that actually matter.

**Review by jslobodzian — CHANGES_REQUESTED** [review]
Can we narrow the change to just the impacted flags.

File: SPECS/python-optree/python-optree.spec (hunk 2 above)
Comment (PawelWMS) [reply SPECS/python-optree/python-optree.spec:39]: +1 to doing only surgical changes to the flags we actually need to change. Otherwise, the specs (or any piece of code, really) gets noisy, it's hard to understand the true intention and which changes we actually need. In general, never include redundant/unnecessary lines. It looks like you're very close now.

File: SPECS/python-optree/python-optree.spec [hunk 3]
```diff
@@ -30,18 +32,28 @@ A PyTree is a recursive structure that can be an arbitrarily nested Python conta
 
 
 %build
+
+export FFLAGS=$(echo $FFLAGS | sed 's/\(-Wp,-D_GLIBCXX_ASSERTIONS\)//g')
```
Comment (PawelWMS) [inline SPECS/python-optree/python-optree.spec:36]: Aren't we're missing the `%check` section? I understood we caught the issue when you were adding a simple tests, which tried to initialize a tree with an empty tuple.

File: SPECS/keras/0001-Add-Keras-3.3.3.patch (hunk 1 above)
Comment (rikenm1) [reply SPECS/keras/0001-Add-Keras-3.3.3.patch:17]: The error was it would complain about missing tensorflow module.  Tensorflow used to build time dep but with Keras 3.0 not anymore. Keras also doesn't use bazel anymore either.  
Currently, TensorFlow is not even runtime nor a build time dependency of keras. Keras is runtime req of tensorflow.  "pip3 install keras" doesn't install tensorflow.  Hence hardcoded this. If not it would have created TensorFlow as a build time req just to get the version number. 
I will add a variable.

File: SPECS/python-optree/python-optree.spec (hunk 3 above)
Comment (rikenm1) [reply SPECS/python-optree/python-optree.spec:36]: We caught it from the smoke test for tensorflow container. when running %pytest, it complains about missing helper files.  Looks like test provided by python-optree doesn't work properly.

File: SPECS/keras/0001-Add-Keras-3.3.3.patch (hunk 1 above)
Comment (rikenm1) [reply SPECS/keras/0001-Add-Keras-3.3.3.patch:17]: here's the stack trace. 
image

File: SPECS/python-optree/python-optree.spec (hunk 2 above)
Comment (rikenm1) [reply SPECS/python-optree/python-optree.spec]: updated

File: SPECS/python-optree/python-optree.spec [hunk 4]
```diff
@@ -30,18 +32,24 @@ A PyTree is a recursive structure that can be an arbitrarily nested Python conta
 
 
 %build
+# Remove -D_GLIBCXX_ASSERTIONS because optree throws std lib assertion errors.
```
Comment (PawelWMS) [inline SPECS/python-optree/python-optree.spec:35]: Sounds a bit like we're burying a legitimate error, which is not really what's happening. How about this:
```suggestion
# Remove "-D_GLIBCXX_ASSERTIONS" because optree will otherwise crash
# on an C++ std libs assertion error. This happens when one tries to
# call 'map_tree' on an empty tuple (or possibly all empty collections).
# We currently suspect an upstream bug, as 'map_tree' seems to work
# fine when compiled without the "_GLIBCXX_ASSERTIONS" macro.
```

**Review by PawelWMS — APPROVED** [review]

**Review by jslobodzian — APPROVED** [review]
